tensorboard/components/vz_line_chart2/vz-line-chart2.ts

```diff
@@ -1,454 +1,393 @@
-       * - "right" - Position the tooltip to the right of the chart.
-       * - "auto" - Position the tooltip to the bottom of the chart in most case.
-       *            Position the tooltip above the chart if there isn't sufficient
-       *            space below.
-       */
-      tooltipPosition: {
-        type: String,
-        value: vz_chart_helper.TooltipPosition.BOTTOM,
-      },
-
-      _chart: Object,
-      _visibleSeriesCache: {
-        type: Array,
-        value: () => [],
-      },
-      _seriesDataCache: {
-        type: Object,
-        value: () => ({}),
-      },
-      _seriesMetadataCache: {
-        type: Object,
-        value: () => ({}),
-      },
-      _makeChartAsyncCallbackId: {type: Number, value: null},
-    },
-
-    observers: [
-      '_makeChart(xComponentsCreationMethod, xType, yValueAccessor, yScaleType, tooltipColumns, colorScale, isAttached)',
-      // Refer to the cache and, if available, load data of a new visible series.
-      '_reloadFromCache(_chart, _visibleSeriesCache)',
-      '_smoothingChanged(smoothingEnabled, smoothingWeight, _chart)',
-      '_tooltipSortingMethodChanged(tooltipSortingMethod, _chart)',
-      '_outliersChanged(ignoreYOutliers, _chart)',
-    ],
-
-    ready() {
-      this.scopeSubtree(this.$.chartdiv, true);
-    },
-
-    attached() {
-      // `capture` ensures that no handler can stop propagation and break the
-      // handler. `passive` ensures that browser does not wait renderer thread
-      // on JS handler (which can prevent default and impact rendering).
-      const option = {capture: true, passive: true};
-      this._listen(this, 'mousedown', this._onMouseDown.bind(this), option);
-      this._listen(this, 'mouseup', this._onMouseUp.bind(this), option);
-      this._listen(window, 'keydown', this._onKeyDown.bind(this), option);
-      this._listen(window, 'keyup', this._onKeyUp.bind(this), option);
-    },
-
-    detached() {
-      this.cancelAsync(this._makeChartAsyncCallbackId);
-      if (this._chart) this._chart.destroy();
-      if (this._listeners) {
-        this._listeners.forEach(({node, eventName, func, option}) => {
-          node.removeEventListener(eventName, func, option);
-        });
-        this._listeners.clear();
+  },
+  {
+    title: 'Value',
+    evaluate: (d: vz_chart_helpers.Point) => formatValueOrNaN(d.datum.scalar),
+  },
+  {
+    title: 'Step',
+    evaluate: (d: vz_chart_helpers.Point) =>
+      vz_chart_helpers.stepFormatter(d.datum.step),
+  },
+  {
+    title: 'Time',
+    evaluate: (d: vz_chart_helpers.Point) =>
+      vz_chart_helpers.timeFormatter(d.datum.wall_time),
+  },
+  {
+    title: 'Relative',
+    evaluate: (d: vz_chart_helpers.Point) =>
+      vz_chart_helpers.relativeFormatter(
+        vz_chart_helpers.relativeAccessor(d.datum, -1, d.dataset)
+      ),
+  },
+];
+
+type NumberFormatter = (num: number) => string;
+
+type EventAttachable = Node | Window;
```

Comment:
Maybe just `EventTarget`, instead of creating a new type?

Reply:
Didn't know about this. Thanks!